app/src/main/java/nz/org/cacophony/birdmonitor/PermissionsHelper.java

```diff
@@ -0,0 +1,84 @@
+package nz.org.cacophony.birdmonitor;
+// https://stackoverflow.com/questions/34342816/android-6-0-multiple-permissions
+
+import android.app.Activity;
+import android.app.AlertDialog;
+import android.content.DialogInterface;
+import android.content.pm.PackageManager;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PermissionsHelper {
+
+    private static final int REQUEST_ID_MULTIPLE_PERMISSIONS = 100; // any code you want.
+
+    public void checkAndRequestPermissions(Activity activity, String... permissions) {
+        List<String> listPermissionsNeeded = new ArrayList<>();
+        for (String permission : permissions) {
+            if (ContextCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
+                listPermissionsNeeded.add(permission);
+            }
+        }
+        if (!listPermissionsNeeded.isEmpty()) {
+            ActivityCompat.requestPermissions(activity, listPermissionsNeeded.toArray(new String[listPermissionsNeeded.size()]), REQUEST_ID_MULTIPLE_PERMISSIONS);
+        }
+    }
+
+    public void onRequestPermissionsResult(Activity activity, int requestCode, String permissions[], int[] grantResults) {
+        switch (requestCode) {
+            case REQUEST_ID_MULTIPLE_PERMISSIONS: {
+                Map<String, Integer> perms = new HashMap<>();
+
+                for (String permission : permissions) {
+                    perms.put(permission, PackageManager.PERMISSION_GRANTED);
+                }
+
+                if (grantResults.length > 0) {
+                    for (int i = 0; i < permissions.length; i++)
+                        perms.put(permissions[i], grantResults[i]);
+
+                    boolean allPermissionsGranted = true;
+                    for (String permission1 : permissions) {
+                        allPermissionsGranted = allPermissionsGranted && (perms.get(permission1) == PackageManager.PERMISSION_GRANTED);
+                    }
+
+                    if (allPermissionsGranted) {
+                        Log.d(PermissionsHelper.class.getSimpleName(), "onRequestPermissionsResult: all permissions granted");
+                    } else {
+                        for (String permission2 : perms.keySet())
+                            if (perms.get(permission2) == PackageManager.PERMISSION_GRANTED)
+                                perms.remove(permission2);
```

Comment:
The above looks more complicated than it needs to be. Why not just loop over `grantResults` and append `permissions[i]` to a `String[]` for each each `PERMISSION_DENIED` value? If the resulting array has a length > 0 then not all permissions were granted and the array can then be used to generate the message to the user.

There's no need for the `perms` hashmap, the `allPermissionsGranted` variable or the multiple loops.

Reply:
done - but slightly more complicated as grantResults is non empty when a user has originally denied a permission and then later grants a permission.  Just need to check the the values:

grantResults[i] != PackageManager.PERMISSION_GRANTED